### ISSUE TITLE
Passenger + Apache http 400 errors - max request headers issue

### ISSUE THREAD

**Issue by bryantcj52** [phusion/passenger#2260]
We are switching from Passenger Standalone to Passenger + Apache for one of our apps and we're having issues with http 400 errors. I've narrowed it down request header size but with numbers that don't make any sense.

Using the same app on two servers, one with Passenger Standalone/Nginx and one with Passenger + Apache:

**Standalone:**
passenger_buffer_size 16k;
passenger_buffers 8 16k;
proxy_buffer_size 16k;
proxy_buffers 8 16k;
large_client_header_buffers 4 16k;

Requests with roughly 2875 bytes of headers works as expected.

**Apache:**
LimitRequestLine 16384
LimitRequestFieldSize 16384

Requests going anywhere in apache non-passenger requests with 2875 bytes of headers work fine (we have other vhosts and requests to the same vhost that runs passenger but hits the public directory and bypasses passenger). As soon as we hit passenger with 2875 bytes, it returns a 400. Removing a couple cookies down to 2586 bytes of headers works fine. The request is stopped by passenger before it even hits the app.

Apache default is 8kb as is node. Running the same app code against nginx works fine.

Turning on passenger debug logging to max produces no info:
> [ D 2020-03-16 12:52:28.8934 21812/T8 age/Cor/Con/TurboCaching.h:246 ]: Clearing turbocache
> [ D 2020-03-16 12:52:28.8935 21812/T8 Ser/Server.h:880 ]: [ServerThr.1] 1 new client(s) accepted; there are now 1 active client(s)
[ D 2020-03-16 12:52:28.8937 21812/T8 age/Cor/App/Gro/SessionManagement.cpp:330 ]: Session checked out from process (pid=21973, group=/local/apps/rbf/web-checkin (production))
[ D 2020-03-16 12:52:28.8937 21812/T8 age/Cor/Con/CheckoutSession.cpp:139 ]: [Client 1-4] Session checked out: pid=21973, gupid=192ed78-kgCJChF6h2
[ D 2020-03-16 12:52:28.8938 21812/T8 age/Cor/Con/CheckoutSession.cpp:199 ]: [Client 1-4] Session initiated: fd=53
[ D 2020-03-16 12:52:28.8944 21812/T8 age/Cor/Con/ForwardResponse.cpp:423 ]: [Client 1-4] Processing turbocache invalidation based on response

Apache/2.4.6 (CentOS) OpenSSL/1.0.2k-fips Phusion_Passenger/6.0.4
Centos 7 3.10.0-1062.9.1.el7.x86_64 
Phusion YUM repo
Nodejs 10.x

**Comment by CamJN** [comment]
After testing this quite a bit, I don't _think_ this is a passenger issue, I can process requests with total headers up to 8KB on my box without changing my apache conf and up to 16KB by setting the following at the top level of my `httpd.conf`:

`LimitRequestFieldSize 16384`

Note: Setting this config in a virtual host behaves in a really unexpected way, per: https://httpd.apache.org/docs/2.4/mod/core.html#limitrequestfieldsize

> When name-based virtual hosting is used, the value for this directive is taken from the default (first-listed) virtual host best matching the current IP address and port combination.

So if you have more than one virtual host, and they all listen on the same port/ip (the most normal apache configuration by far) then the first virtualhost determines the limits for ALL virtual hosts. You can figure out which virtual host is considered "first" with `apachectl -S`. But my recommendation is to increase the limits in a higher level than the virtualhost level.

**Comment by bryantcj52** [comment]
We ended up having to have apache proxy back to the app running in standalone/builtin mode.

I tried again using mod_passenger 6.0.7 and instead of a 400, I'm getting 431 Request Header Fields Too Large with a request only 3572 in size, well below the default 8kb. Cutting the cookie header down in half, the request worked. I tried adding the two limit lines to the default vhost and the vhost running the app, no luck.
